Xamarin.Forms.Core/Messaging.cs

```diff
@@ -0,0 +1,158 @@
+			{
+				var send = (TSender)sender;
+				if (source == null || send == source)
+					callback((TSender)sender, (TArgs)args);
+			};
+
+			InnerSubscribe(subscriber, message, typeof(TSender), typeof(TArgs), wrap);
+		}
+
+		public void Subscribe<TSender>(object subscriber, string message, Action<TSender> callback, TSender source = null) where TSender : class
+		{
+			if (subscriber == null)
+				throw new ArgumentNullException(nameof(subscriber));
+
+			if (string.IsNullOrWhiteSpace(message))
+				throw new ArgumentException(nameof(message) + " name cannot be empty.");
+
+			if (callback == null)
+				throw new ArgumentNullException(nameof(callback));
+
+			Action<object, object> wrap = (sender, args) =>
+			{
+				var send = (TSender)sender;
+				if (source == null || send == source)
+					callback((TSender)sender);
+			};
+
+			InnerSubscribe(subscriber, message, typeof(TSender), null, wrap);
+		}
+
+		public void Unsubscribe<TSender, TArgs>(object subscriber, string message) where TSender : class
+		{
+			if (subscriber == null)
+				throw new ArgumentNullException(nameof(subscriber));
+
+			if (string.IsNullOrWhiteSpace(message))
+				throw new ArgumentException(nameof(message) + " name cannot be empty.");
+
+			InnerUnsubscribe(message, typeof(TSender), typeof(TArgs), subscriber);
+		}
+
+		public void Unsubscribe<TSender>(object subscriber, string message) where TSender : class
+		{
+			if (subscriber == null)
+				throw new ArgumentNullException(nameof(subscriber));
+
+			if (string.IsNullOrWhiteSpace(message))
+				throw new ArgumentException(nameof(message) + " name cannot be empty.");
+
+			InnerUnsubscribe(message, typeof(TSender), null, subscriber);
+		}
+
+		internal void ClearSubscribers()
+		{
+			_callbacks.Clear();
+		}
+
+		void InnerSend(string message, Type senderType, Type argType, object sender, object args)
+		{
+			var key = new Tuple<string, Type, Type>(message, senderType, argType);
+
+			if (!_callbacks.ContainsKey(key))
+				return;
+
+			List<Tuple<WeakReference, Action<object, object>>> actions = _callbacks[key];
+			if (actions == null || !actions.Any())
+				return; // should not be reachable
+
+			// ok so this code looks a bit funky but here is the gist of the problem. It is possible that in the course
+			// of executing the callbacks for this message someone will subscribe/unsubscribe from the same message in
+			// the callback. This would invalidate the enumerator. To work around this we make a copy. However if you unsubscribe 
+			// from a message you can fairly reasonably expect that you will therefore not receive a call. To fix this we then
+			// check that the item we are about to send the message to actually exists in the live list.
+			List<Tuple<WeakReference, Action<object, object>>> actionsCopy = actions.ToList();
+			foreach (Tuple<WeakReference, Action<object, object>> action in actionsCopy)
+			{
+				if (action.Item1.Target != null && actions.Contains(action))
```

Comment:
Proper way to check if an object is alive. Created a bug: https://bugzilla.xamarin.com/show_bug.cgi?id=47203